src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java

```diff
@@ -112,6 +106,9 @@ public void actionPerformed(@NotNull AnActionEvent e) {
         ideRestarter.restart();
       }
     }
+
+    MainViewModelUpdater mainViewModelUpdater = new MainViewModelUpdater(project);
+    new Thread(mainViewModelUpdater::run).start();
```

Comment:
Java threads are on quite low level of abstraction. Did you take a look if IntelliJ platform provides some "background worker" objects to do things on background. On the other hand, as this thread is forever-running, there's no need for thread-pooling and such. But still this makes me a bit nervous. :)

Btw, is this thread (silently) terminated, if the plugin is disabled/uninstalled?

Reply:
Well it's not strictly forever-running, since it terminates if the project associated with it gets closed. But yes, I agree that Java threads are very low level. I did find IntelliJ's `ProgressManager`, but that seemed to have a lot to do with UI related to background processes (e.g. progress bars and the like), but I'll check again if IntelliJ has some convenient abstractions for this.

I overlooked plugin uninstallation, so I'll have to test that :smile: